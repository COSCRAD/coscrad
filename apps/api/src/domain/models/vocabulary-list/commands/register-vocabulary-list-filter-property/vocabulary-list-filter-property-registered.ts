import { CoscradEvent } from '../../../../../domain/common';
import { BaseEvent } from '../../../shared/events/base-event.entity';
import { RegisterVocabularyListFilterProperty } from './register-vocabulary-list-filter-property.command';

export type VocabularyListFilterPropertyRegisteredPayload = RegisterVocabularyListFilterProperty;

@CoscradEvent('VOCABULARY_LIST_PROPERTY_FILTER_REGISTERED')
export class VocabularyListFilterPropertyRegistered extends BaseEvent<VocabularyListFilterPropertyRegisteredPayload> {
    readonly type = 'VOCABULARY_LIST_PROPERTY_FILTER_REGISTERED';
}
