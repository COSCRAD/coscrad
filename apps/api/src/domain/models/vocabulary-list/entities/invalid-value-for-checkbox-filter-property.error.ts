import { InternalError } from '../../../../lib/errors/InternalError';
import { ValueAndDisplay } from './value-and-display.entity';

export class InvalidValueForCheckboxFilterPropertyError extends InternalError {
    constructor({ value, display }: ValueAndDisplay<unknown>, name: string) {
        const msg = `You cannot add the value: ${value} (${display}) as the value of the checkbox vocabulary list filter property ${name}.. The only values that are permitted for checkbox filters are true or false.`;

        super(msg);
    }
}
