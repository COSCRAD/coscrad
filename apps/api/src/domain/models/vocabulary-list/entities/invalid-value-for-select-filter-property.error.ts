import { InternalError } from '../../../../lib/errors/InternalError';
import { ValueAndDisplay } from './value-and-display.entity';

export class InvalidValueForSelectFilterPropertyError extends InternalError {
    constructor({ value, display: display }: ValueAndDisplay<unknown>, name: string) {
        const msg = `You cannot add the value: ${value} (${display}) as the value of the select vocabulary list filter property ${name}.. Only string values are permitted for select filters.`;

        super(msg);
    }
}
