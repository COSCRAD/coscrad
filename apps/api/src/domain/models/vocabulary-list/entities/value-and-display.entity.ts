import { IValueAndDisplay } from '@coscrad/api-interfaces';
import { ApiProperty } from '@nestjs/swagger';

export class ValueAndDisplay<T> implements IValueAndDisplay<T> {
    @ApiProperty({
        example: '11',
        description: 'a string or boolean value to be used in filtering a vocabulary list',
    })
    public readonly value: T;

    @ApiProperty({
        example: 'I',
        description: 'the user-facing label for the given value',
    })
    public readonly display: string;
}
