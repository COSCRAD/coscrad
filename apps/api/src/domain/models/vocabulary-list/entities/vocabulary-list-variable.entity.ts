import { NonEmptyString } from '@coscrad/data-types';
import { ApiProperty } from '@nestjs/swagger';
import { DropboxOrCheckbox } from '../../vocabulary-list/types/dropbox-or-checkbox';
import { VocabularyListVariableValue } from '../../vocabulary-list/types/vocabulary-list-variable-value';
import { ValueAndDisplay } from './value-and-display.entity';

/**
 * TODO correlate `type` and `validValues`
 * type:dropbox <-> string values
 * type:checkbox <-> boolean values
 */
export class VocabularyListVariable<
    TVariableType extends VocabularyListVariableValue = VocabularyListVariableValue
> {
    @ApiProperty({
        example: 'person',
        description: 'name of a property that parametrizes terms in the list',
    })
    @NonEmptyString()
    name: string;

    @ApiProperty({
        example: 'dropbox',
        description:
            'specifies whether the corresponding field be a dropbox (select) or slider (switch)',
    })
    // TODO Support `DropboxOrCheckbox` enum data-type
    type: DropboxOrCheckbox;

    @ApiProperty({
        description: 'specifies the value and label for the corresponding form element',
    })
    // TODO Add data-type decorator
    validValues: ValueAndDisplay<TVariableType>[];
}
