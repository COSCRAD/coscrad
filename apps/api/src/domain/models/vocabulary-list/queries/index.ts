export * from './vocabulary-list-query-repository.interface';
