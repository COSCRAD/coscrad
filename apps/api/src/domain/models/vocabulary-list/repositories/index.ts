export * from './arango-vocabulary-list-query-repository';
