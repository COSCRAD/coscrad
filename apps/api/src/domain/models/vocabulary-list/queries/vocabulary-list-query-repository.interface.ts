import {
    IDetailQueryResult,
    IMultilingualTextItem,
    IVocabularyListViewModel,
} from '@coscrad/api-interfaces';
import { Maybe } from '../../../../lib/types/maybe';
import { AggregateId } from '../../../types/AggregateId';
import { FilterPropertyType } from '../commands';
import { VocabularyListEntryImportItem } from '../entities/vocabulary-list.entity';

export const VOCABULARY_LIST_QUERY_REPOSITORY_TOKEN = 'VOCABULARY_LIST_QUERY_REPOSITORY_TOKEN';

/**
 * TODO Absorb the detail query response piece into the view model interface.
 * However, there may be a case that these interfaces are actually the types
 * of the query response, as they represent the contract with the client. There
 * are cases where the query service is responsible for mapping the view model further,
 * e.g. by appending a base URL or leveraging the media item or user services, for example.
 * We should sort this out before we do a fourth event-sourced query service.
 */
type VocabularyListQueryModel = IDetailQueryResult<IVocabularyListViewModel>;

export interface IVocabularyListQueryRepository {
    fetchById(id: AggregateId): Promise<Maybe<VocabularyListQueryModel>>;

    fetchMany(): Promise<VocabularyListQueryModel[]>;

    count(): Promise<number>;

    create(view: VocabularyListQueryModel): Promise<void>;

    createMany(views: VocabularyListQueryModel[]): Promise<void>;

    delete(id: AggregateId): Promise<void>;

    allowUser(id: AggregateId, userId: AggregateId): Promise<void>;

    // translateName(id: AggregateId, translationItem: IMultilingualTextItem): Promise<void>;

    publish(id: AggregateId): Promise<void>;

    /**
     * A better approach would be to do this atomically as part of
     * each update query. We need to find a performant and extensible way to
     * do this.
     */
    attribute(id: AggregateId, contributorIds: AggregateId[]): Promise<void>;

    translateName(id: AggregateId, translationItem: IMultilingualTextItem): Promise<void>;

    registerFilterProperty(
        id: AggregateId,
        name: string,
        type: FilterPropertyType,
        allowedValuesAndLabels: { value: string | boolean; label: string }[]
    ): Promise<void>;

    addTerm(vocabularyListId: AggregateId, termId: AggregateId): Promise<void>;

    analyzeTerm(
        vocabularyListId: AggregateId,
        termId: AggregateId,
        propertyValues: Record<string, string | boolean>
    ): Promise<void>;

    importEntries(
        vocabularyListId: AggregateId,
        entries: VocabularyListEntryImportItem[]
    ): Promise<void>;
}
