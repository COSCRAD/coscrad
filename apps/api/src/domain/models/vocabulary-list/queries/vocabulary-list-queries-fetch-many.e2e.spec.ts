import {
    AggregateType,
    IIndexQueryResult,
    IMultilingualTextItem,
    IVocabularyListViewModel,
    LanguageCode,
} from '@coscrad/api-interfaces';
import { INestApplication } from '@nestjs/common';
import * as request from 'supertest';
import httpStatusCodes, { HttpStatusCode } from '../../../../app/constants/httpStatusCodes';
import setUpIntegrationTest from '../../../../app/controllers/__tests__/setUpIntegrationTest';
import { ConsoleCoscradCliLogger } from '../../../../coscrad-cli/logging';
import { isNotFound } from '../../../../lib/types/not-found';
import { ArangoConnectionProvider } from '../../../../persistence/database/arango-connection.provider';
import { ArangoDatabaseProvider } from '../../../../persistence/database/database.provider';
import TestRepositoryProvider from '../../../../persistence/repositories/__tests__/TestRepositoryProvider';
import generateDatabaseNameForTestSuite from '../../../../persistence/repositories/__tests__/generateDatabaseNameForTestSuite';
import { TermViewModel } from '../../../../queries/buildViewModelForResource/viewModels/term.view-model';
import { TestEventStream } from '../../../../test-data/events';
import { getCoscradEventConsumerMeta, ICoscradEvent } from '../../../common';
import { buildMultilingualTextFromBilingualText } from '../../../common/build-multilingual-text-from-bilingual-text';
import { MultilingualText } from '../../../common/entities/multilingual-text';
import { SyncInMemoryEventPublisher } from '../../../common/events/sync-in-memory-event-publisher';
import buildDummyUuid from '../../__tests__/utilities/buildDummyUuid';
import { IAudioItemQueryRepository } from '../../audio-visual/audio-item/queries/audio-item-query-repository.interface';
import { ArangoAudioItemQueryRepository } from '../../audio-visual/audio-item/repositories/arango-audio-item-query-repository';
import { AccessControlList } from '../../shared/access-control/access-control-list.entity';
import { ResourcePublished } from '../../shared/common-commands/publish-resource/resource-published.event';
import { ResourcePublishedEventHandler } from '../../shared/common-commands/publish-resource/resource-published.event-handler';
import { ITermQueryRepository } from '../../term/queries';
import { ArangoTermQueryRepository } from '../../term/repositories';
import { ArangoQueryRepositoryProvider } from '../../term/repositories/arango-query-repository-provider';
import {
    TermAddedToVocabularyList,
    TermInVocabularyListAnalyzed,
    VocabularyListCreated,
    VocabularyListCreatedEventHandler,
    VocabularyListFilterPropertyRegistered,
} from '../commands';
import { TermAddedToVocabularyListEventHandler } from '../commands/add-term-to-vocabulary-list/term-added-to-vocabulary-list.event-handler';
import { TermInVocabularyListAnalyzedEventHandler } from '../commands/analyze-term-in-vocabulary-list/term-in-vocabulary-list-analyzed.event-handler';
import { VocabularyListFilterPropertyRegisteredEventHandler } from '../commands/register-vocabulary-list-filter-property/vocabulary-list-filter-property-registered.event-handler';
import { VocabularyListNameTranslated } from '../commands/translate-vocabulary-list-name/vocabulary-list-name-translated.event';
import { VocabularyListNameTranslatedEventHandler } from '../commands/translate-vocabulary-list-name/vocabulary-list-name-translated.event-handler';
import { ArangoVocabularyListQueryRepository } from '../repositories';
import { IVocabularyListQueryRepository } from './vocabulary-list-query-repository.interface';

const vocabularyListId = buildDummyUuid(4);

const vocabularyListName = 'to swim';

const languageCodeForVocabularyListName = LanguageCode.English;

const termId = buildDummyUuid(5);

const termOriginalText = 'I swim';

const termOriginalLanguageCode = LanguageCode.Chilcotin;

const termTranslation = 'I am swimming (English)';

const termTranslationLanguageCode = LanguageCode.English;

const indexEndpoint = `/resources/vocabularyLists`;

/**
 * TODO We may want to event source our test setup to give us a very high
 * level of confidence that our back-end is working from event handler
 * to query endpoint. Eagerly joining in terms, tags, notes, etc. into vocabulary
 * lists, for example, is a big risk.
 */
const existingTerm: TermViewModel = TermViewModel.fromDto({
    id: termId,
    isPublished: true,
    accessControlList: new AccessControlList().toDTO(),
    actions: [],
    name: buildMultilingualTextFromBilingualText(
        {
            text: termOriginalText,
            languageCode: termOriginalLanguageCode,
        },
        {
            text: termTranslation,
            languageCode: termTranslationLanguageCode,
        }
    ),
    contributions: [],
});

const vocabularyListEventStream = new TestEventStream()
    .andThen<VocabularyListCreated>({
        type: 'VOCABULARY_LIST_CREATED',
        payload: {
            name: vocabularyListName,
            languageCodeForName: languageCodeForVocabularyListName,
        },
    })
    .andThen<VocabularyListNameTranslated>({
        type: 'VOCABULARY_LIST_NAME_TRANSLATED',
    })
    .andThen<VocabularyListFilterPropertyRegistered>({
        type: 'VOCABULARY_LIST_PROPERTY_FILTER_REGISTERED',
    })
    .andThen<TermAddedToVocabularyList>({
        type: 'TERM_ADDED_TO_VOCABULARY_LIST',
        payload: {
            termId,
        },
    })
    .andThen<TermInVocabularyListAnalyzed>({
        type: 'TERM_IN_VOCABULARY_LIST_ANALYZED',
    })
    .andThen<ResourcePublished>({
        type: 'RESOURCE_PUBLISHED',
    });

// TODO check that contributors are joined in
// const existingContributor = getValidAggregateInstanceForTest(AggregateType.contributor)

const eventPublisher = new SyncInMemoryEventPublisher(new ConsoleCoscradCliLogger());

const registerEventHandlers = (
    vocabularyListQueryRepository: IVocabularyListQueryRepository,
    termQueryRepository: ITermQueryRepository,
    audioItemQueryRepository: IAudioItemQueryRepository
): void => {
    const specificCtors = [
        VocabularyListCreatedEventHandler,
        VocabularyListNameTranslatedEventHandler,
        VocabularyListFilterPropertyRegisteredEventHandler,
        TermAddedToVocabularyListEventHandler,
        TermInVocabularyListAnalyzedEventHandler,
    ];

    const genericCtors = [ResourcePublishedEventHandler];

    const handlers = [];

    /**
     * We should use the test module to set this up.
     *
     * We are trying very hard to avoid introducing the handlers into
     * `setUpIntegrationTest` \ `createTestModule`, which already have been
     * identified as an anti-pattern that hurts extesnibility.
     */
    specificCtors.forEach((Ctor) => {
        const meta = getCoscradEventConsumerMeta(Ctor);

        if (isNotFound(meta)) {
            return;
        }

        const { type: eventType } = meta;

        const handler = new Ctor(vocabularyListQueryRepository);

        eventPublisher.register(eventType, handler);

        handlers.push(handler);
    });

    genericCtors.forEach((Ctor) => {
        const meta = getCoscradEventConsumerMeta(Ctor);

        if (isNotFound(meta)) {
            return;
        }

        const { type: eventType } = meta;

        const handler = new Ctor(
            new ArangoQueryRepositoryProvider(
                termQueryRepository,
                audioItemQueryRepository,
                vocabularyListQueryRepository
            )
        );

        eventPublisher.register(eventType, handler);

        handlers.push(handler);
    });
};

/**
 * TODO[test-coverage]: include test case with import history
 *
 * THIS TEST IS BEING REWRITTEN ON ANOTHER BRANCH. OPT BACK IN IMMEDIATELY WHEN REBASING.
 *
 * **
 */

describe.skip(`when querying for a vocabulary list: fetch many`, () => {
    const testDatabaseName = generateDatabaseNameForTestSuite();

    let app: INestApplication;

    let testRepositoryProvider: TestRepositoryProvider;

    let databaseProvider: ArangoDatabaseProvider;

    let connectionProvider: ArangoConnectionProvider;

    let vocabularyListQueryRepository: IVocabularyListQueryRepository;

    let seedVocabularyLists: (eventHistory: ICoscradEvent[]) => Promise<void>;

    beforeEach(async () => {
        await testRepositoryProvider.testSetup();

        await databaseProvider.getDatabaseForCollection('term__VIEWS').clear();

        await databaseProvider.getDatabaseForCollection('vocabularyList__VIEWS').clear();
    });

    afterAll(async () => {
        await app.close();

        databaseProvider.close();
    });

    describe(`when the user is unauthenticated`, () => {
        beforeAll(async () => {
            // TODO avoid using `setUpIntegrationTest` here
            ({ app, testRepositoryProvider, databaseProvider } = await setUpIntegrationTest(
                {
                    ARANGO_DB_NAME: testDatabaseName,
                }
                // no authenticated user
            ));

            connectionProvider = app.get(ArangoConnectionProvider);

            vocabularyListQueryRepository = new ArangoVocabularyListQueryRepository(
                connectionProvider,
                new ConsoleCoscradCliLogger()
            );

            connectionProvider = app.get(ArangoConnectionProvider);

            const audioItemRepo = new ArangoAudioItemQueryRepository(connectionProvider);

            registerEventHandlers(
                vocabularyListQueryRepository,
                new ArangoTermQueryRepository(
                    connectionProvider,
                    audioItemRepo,
                    new ConsoleCoscradCliLogger()
                ),
                audioItemRepo
            );

            seedVocabularyLists = async (events: ICoscradEvent[]) => {
                for (const e of events) {
                    await eventPublisher.publish(e);
                }
            };
        });

        describe(`when there is a vocabulary list with the given ID`, () => {
            describe(`when the vocabulary list is published`, () => {
                beforeEach(async () => {
                    const eventHistory = vocabularyListEventStream.as({
                        type: AggregateType.vocabularyList,
                        id: vocabularyListId,
                    });

                    await new ArangoTermQueryRepository(
                        connectionProvider,
                        new ArangoAudioItemQueryRepository(connectionProvider),
                        new ConsoleCoscradCliLogger()
                    ).create(existingTerm);

                    await seedVocabularyLists(eventHistory);
                });

                it(`should return the correct vocabulary list view`, async () => {
                    const res = await request(app.getHttpServer()).get(indexEndpoint);

                    expect(res.status).toBe(HttpStatusCode.ok);

                    const body = res.body as IIndexQueryResult<IVocabularyListViewModel>;

                    const { entities: foundEntities } = body;

                    expect(foundEntities).toHaveLength(1);

                    const { name: namedto, entries } = foundEntities[0];

                    const foundName = new MultilingualText(namedto);

                    const { text: foundText, languageCode: foundLanguageCode } =
                        foundName.getOriginalTextItem();

                    expect(foundText).toBe(vocabularyListName);

                    expect(foundLanguageCode).toBe(languageCodeForVocabularyListName);

                    // here we check that the term comes through
                    expect(entries).toHaveLength(1);

                    const { term: foundTerm } = entries[0];

                    const { name: termNameDto } = foundTerm;

                    const termName = new MultilingualText(termNameDto);

                    const { text: foundTermText, languageCode: foundTermLanguageCode } =
                        termName.getOriginalTextItem();

                    expect(foundTermText).toBe(termOriginalText);

                    expect(foundTermLanguageCode).toBe(termOriginalLanguageCode);

                    const {
                        text: foundTermTranslationText,
                        languageCode: foundTermTranslationLanguageCode,
                    } = termName.getTranslation(
                        termTranslationLanguageCode
                    ) as IMultilingualTextItem;

                    expect(foundTermTranslationText).toBe(termTranslation);

                    expect(foundTermTranslationLanguageCode).toBe(termTranslationLanguageCode);
                    // TODO check additional state
                });
            });
        });

        describe(`when there are no vocabulary lists`, () => {
            it(`should return an empty list of entities`, async () => {
                const res = await request(app.getHttpServer()).get(indexEndpoint);

                expect(res.status).toBe(httpStatusCodes.ok);

                expect(res.body.entities).toHaveLength(0);
            });
        });

        describe(`when one of the entries is missing a corresponding term in the database`, () => {
            it.todo(`should fail gracefully`);
        });
    });

    /**
     * Currently, we are pushing to get public assets live in production. When
     * we are ready to start using the system as an admin, we should complete this
     * test coverage.
     */
    describe(`when the user is a COSCRAD admin`, () => {
        it.todo(`should have a test`);
    });

    describe(`when the user is a project admin`, () => {
        it.todo(`should have a test`);
    });

    describe(`when the user is authenticated as an ordinary user`, () => {
        it.todo(`should have a test`);
    });
});
