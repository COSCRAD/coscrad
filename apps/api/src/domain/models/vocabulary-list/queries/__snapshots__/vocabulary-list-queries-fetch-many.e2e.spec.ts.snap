// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when querying for a vocabulary list: fetch many when the user is a COSCRAD admin when some terms are private and some are public should have access to all vocabulary lists and terms 1`] = `
{
  "entities": [
    {
      "actions": [
        {
          "description": "Translate the name of a vocabulary list to another language",
          "form": {
            "fields": [
              {
                "constraints": [
                  {
                    "message": "Must be a valid \${propertyLabel}",
                    "name": "IS_ENUM",
                  },
                  {
                    "message": "Required",
                    "name": "defined value",
                  },
                ],
                "description": "the language in which you are naming the new vocabulary list name",
                "label": "language code",
                "name": "languageCode",
                "options": [
                  {
                    "display": "Chilcotin",
                    "value": "clc",
                  },
                  {
                    "display": "Haida",
                    "value": "hai",
                  },
                  {
                    "display": "English",
                    "value": "en",
                  },
                  {
                    "display": "French",
                    "value": "fra",
                  },
                  {
                    "display": "Chinook",
                    "value": "chn",
                  },
                  {
                    "display": "Zapotec",
                    "value": "zap",
                  },
                  {
                    "display": "Spanish",
                    "value": "spa",
                  },
                ],
                "type": "STATIC_SELECT",
              },
              {
                "constraints": [
                  {
                    "message": "must be a non-empty string",
                    "name": "non-empty string",
                  },
                  {
                    "message": "must be a defined value",
                    "name": "defined value",
                  },
                ],
                "description": "text for the translation of the name",
                "label": "translation",
                "name": "text",
                "type": "TEXT_FIELD",
              },
            ],
            "prepopulatedFields": {
              "aggregateCompositeIdentifier": {
                "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b100001",
                "type": "vocabularyList",
              },
            },
          },
          "label": "Translate Vocabulary List Name",
          "type": "TRANSLATE_VOCABULARY_LIST_NAME",
        },
      ],
      "contributions": [
        {
          "contributorIds": [
            "9b1deb4d-3b7d-4bad-9bdd-2b0d7b100374",
          ],
          "date": {
            "day": 27,
            "month": "09",
            "year": 2022,
          },
          "statement": "Watchamacallit fabricated by: Willy Wankie",
          "type": "WHATCHAMACALLIT_FABRICATED",
        },
      ],
      "entries": [
        {
          "term": {
            "accessControlList": {
              "allowedGroupIds": [],
              "allowedUserIds": [],
            },
            "actions": [],
            "contributions": [],
            "id": "123",
            "isPublished": false,
            "mediaItemId": "555",
            "name": {
              "items": [
                {
                  "languageCode": "en",
                  "role": "original",
                  "text": "test term in vocabulary list",
                },
              ],
            },
            "vocabularyLists": [],
          },
          "variableValues": {},
        },
      ],
      "form": {
        "fields": [],
      },
      "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b100001",
      "isPublished": true,
      "name": {
        "items": [
          {
            "languageCode": "en",
            "role": "original",
            "text": "to swim",
          },
          {
            "languageCode": "clc",
            "role": "free translation",
            "text": "to swim (translation)",
          },
        ],
      },
    },
    {
      "actions": [
        {
          "description": "Translate the name of a vocabulary list to another language",
          "form": {
            "fields": [
              {
                "constraints": [
                  {
                    "message": "Must be a valid \${propertyLabel}",
                    "name": "IS_ENUM",
                  },
                  {
                    "message": "Required",
                    "name": "defined value",
                  },
                ],
                "description": "the language in which you are naming the new vocabulary list name",
                "label": "language code",
                "name": "languageCode",
                "options": [
                  {
                    "display": "Chilcotin",
                    "value": "clc",
                  },
                  {
                    "display": "Haida",
                    "value": "hai",
                  },
                  {
                    "display": "English",
                    "value": "en",
                  },
                  {
                    "display": "French",
                    "value": "fra",
                  },
                  {
                    "display": "Chinook",
                    "value": "chn",
                  },
                  {
                    "display": "Zapotec",
                    "value": "zap",
                  },
                  {
                    "display": "Spanish",
                    "value": "spa",
                  },
                ],
                "type": "STATIC_SELECT",
              },
              {
                "constraints": [
                  {
                    "message": "must be a non-empty string",
                    "name": "non-empty string",
                  },
                  {
                    "message": "must be a defined value",
                    "name": "defined value",
                  },
                ],
                "description": "text for the translation of the name",
                "label": "translation",
                "name": "text",
                "type": "TEXT_FIELD",
              },
            ],
            "prepopulatedFields": {
              "aggregateCompositeIdentifier": {
                "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b100003",
                "type": "vocabularyList",
              },
            },
          },
          "label": "Translate Vocabulary List Name",
          "type": "TRANSLATE_VOCABULARY_LIST_NAME",
        },
      ],
      "contributions": [],
      "entries": [
        {
          "term": {
            "accessControlList": {
              "allowedGroupIds": [],
              "allowedUserIds": [
                "system4d-3b7d-4bad-9bdd-2b0d70userid",
              ],
            },
            "actions": [],
            "contributions": [],
            "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b100103",
            "isPublished": false,
            "name": {
              "items": [
                {
                  "languageCode": "clc",
                  "role": "original",
                  "text": "private term, but with ACL access for user",
                },
              ],
            },
            "vocabularyLists": [],
          },
          "variableValues": {},
        },
      ],
      "form": {
        "fields": [],
      },
      "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b100003",
      "isPublished": false,
      "name": {
        "items": [
          {
            "languageCode": "en",
            "role": "original",
            "text": "vocab list name (orig)",
          },
        ],
      },
    },
    {
      "actions": [
        {
          "description": "Translate the name of a vocabulary list to another language",
          "form": {
            "fields": [
              {
                "constraints": [
                  {
                    "message": "Must be a valid \${propertyLabel}",
                    "name": "IS_ENUM",
                  },
                  {
                    "message": "Required",
                    "name": "defined value",
                  },
                ],
                "description": "the language in which you are naming the new vocabulary list name",
                "label": "language code",
                "name": "languageCode",
                "options": [
                  {
                    "display": "Chilcotin",
                    "value": "clc",
                  },
                  {
                    "display": "Haida",
                    "value": "hai",
                  },
                  {
                    "display": "English",
                    "value": "en",
                  },
                  {
                    "display": "French",
                    "value": "fra",
                  },
                  {
                    "display": "Chinook",
                    "value": "chn",
                  },
                  {
                    "display": "Zapotec",
                    "value": "zap",
                  },
                  {
                    "display": "Spanish",
                    "value": "spa",
                  },
                ],
                "type": "STATIC_SELECT",
              },
              {
                "constraints": [
                  {
                    "message": "must be a non-empty string",
                    "name": "non-empty string",
                  },
                  {
                    "message": "must be a defined value",
                    "name": "defined value",
                  },
                ],
                "description": "text for the translation of the name",
                "label": "translation",
                "name": "text",
                "type": "TEXT_FIELD",
              },
            ],
            "prepopulatedFields": {
              "aggregateCompositeIdentifier": {
                "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b100002",
                "type": "vocabularyList",
              },
            },
          },
          "label": "Translate Vocabulary List Name",
          "type": "TRANSLATE_VOCABULARY_LIST_NAME",
        },
      ],
      "contributions": [],
      "entries": [
        {
          "term": {
            "accessControlList": {
              "allowedGroupIds": [],
              "allowedUserIds": [],
            },
            "actions": [],
            "contributions": [],
            "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b100102",
            "isPublished": false,
            "name": {
              "items": [
                {
                  "languageCode": "clc",
                  "role": "original",
                  "text": "private term",
                },
              ],
            },
            "vocabularyLists": [],
          },
          "variableValues": {},
        },
      ],
      "form": {
        "fields": [],
      },
      "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b100002",
      "isPublished": false,
      "name": {
        "items": [
          {
            "languageCode": "en",
            "role": "original",
            "text": "vocab list name (orig)",
          },
        ],
      },
    },
  ],
  "indexScopedActions": [
    {
      "description": "creates a new vocabulary list",
      "form": {
        "fields": [
          {
            "constraints": [
              {
                "message": "must be a non-empty string",
                "name": "non-empty string",
              },
              {
                "message": "must be a defined value",
                "name": "defined value",
              },
            ],
            "description": "name of the vocabulary list",
            "label": "name",
            "name": "name",
            "type": "TEXT_FIELD",
          },
          {
            "constraints": [
              {
                "message": "Must be a valid \${propertyLabel}",
                "name": "IS_ENUM",
              },
              {
                "message": "Required",
                "name": "defined value",
              },
            ],
            "description": "the language in which you are naming the vocabulary list",
            "label": "language code",
            "name": "languageCodeForName",
            "options": [
              {
                "display": "Chilcotin",
                "value": "clc",
              },
              {
                "display": "Haida",
                "value": "hai",
              },
              {
                "display": "English",
                "value": "en",
              },
              {
                "display": "French",
                "value": "fra",
              },
              {
                "display": "Chinook",
                "value": "chn",
              },
              {
                "display": "Zapotec",
                "value": "zap",
              },
              {
                "display": "Spanish",
                "value": "spa",
              },
            ],
            "type": "STATIC_SELECT",
          },
        ],
        "prepopulatedFields": {},
      },
      "label": "Create Vocabulary List",
      "type": "CREATE_VOCABULARY_LIST",
    },
  ],
}
`;
